apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/DroppedSpanStats.java

```diff
@@ -99,33 +136,39 @@ public void captureDroppedSpan(Span span) {
         stats.sum.addAndGet(span.getDuration());
     }
 
-    private StatsValue getStats(StatsKey statsKey) {
-        StatsValue statsValue = statsMap.get(statsKey);
-        if (statsValue != null) {
-            return statsValue;
+    private Stats getOrCreateStats(StatsKey statsKey) {
```

Comment:
I'm not sure you need the synchronized block, for example I think this works equally without any lock
```
    private Stats getOrCreateStats(StatsKey statsKey) {
        Stats stats = statsMap.get(statsKey);
        if (stats != null || statsMap.size() > 127) {
            statsKeyObjectPool.recycle(statsKey);
            return stats;
        }

        Stats newStats = statsObjectPool.createInstance();
        stats = statsMap.putIfAbsent(statsKey, newStats);
        if (stats == null) {
            return newStats;
        } else {
            statsKeyObjectPool.recycle(statsKey);
            statsObjectPool.recycle(newStats);
            return stats;
        }
    }
```
this doesn't guarantee the 128 limit, but the limit is arbitrary anyway, so exceeding it a little doesn't really matter

Reply:
Yes, Ia also wanted to do something like this, but came to the conclusion that we might exit the limit.

Reply:
I guess most applications won't hit the limit.

Reply:
The limit is an arbitrary number to stop it being something that could grow indefinitely in memory. Just in case the limit is expected on the server too in future (it isn't now), you can limit the JSON write loop to stop at 128

Reply:
so, should we enforce the limit now (when serializing the data)?

Reply:
yes. may as well

Reply:
Done